# Exploratory Data Analysis

Subject: House Prices Prediction

Goal : Predict the sales price for each house in Ames, Iowa. For each Id in the test set, you must predict the value of the SalePrice variable.

Metric : Root-Mean-Squared-Error (RMSE) between the logarithm of the predicted value and the logarithm of the observed sales price. (Taking logs means that errors in predicting expensive houses and cheap houses will affect the result equally.)

## Dataframe details

This dataframe uses US metrics such as square feet.

After much consideration, these metrics will be kept as is because this is a data analysis based on data of the US market and more precisely of the town of Ames in Iowa. It felt more logical to keep the original metric.

But 1 : if it helps, here is how you can translate this data:
1 square meter = 0.09290304 square feet

But 2 : this decision has nothing to do with the fact that I'm lazy.

## Imports

In [1]:
import plotly.express as px
import plotly.graph_objects as go
import polars as pl
from plotly.subplots import make_subplots

PASTEL = px.colors.qualitative.Pastel

## Data analysis

### Size, shape, quantity and type of data

In [2]:
raw = pl.read_parquet("data/train.parquet")
raw.head(10)

Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,LotConfig,LandSlope,Neighborhood,Condition1,Condition2,BldgType,HouseStyle,OverallQual,OverallCond,YearBuilt,YearRemodAdd,RoofStyle,RoofMatl,Exterior1st,Exterior2nd,MasVnrType,MasVnrArea,ExterQual,ExterCond,Foundation,BsmtQual,BsmtCond,BsmtExposure,BsmtFinType1,BsmtFinSF1,BsmtFinType2,BsmtFinSF2,…,2ndFlrSF,LowQualFinSF,GrLivArea,BsmtFullBath,BsmtHalfBath,FullBath,HalfBath,BedroomAbvGr,KitchenAbvGr,KitchenQual,TotRmsAbvGrd,Functional,Fireplaces,FireplaceQu,GarageType,GarageYrBlt,GarageFinish,GarageCars,GarageArea,GarageQual,GarageCond,PavedDrive,WoodDeckSF,OpenPorchSF,EnclosedPorch,3SsnPorch,ScreenPorch,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
i64,i64,str,i64,i64,str,str,str,str,str,str,str,str,str,str,str,str,i64,i64,i64,i64,str,str,str,str,str,i64,str,str,str,str,str,str,str,i64,str,i64,…,i64,i64,i64,i64,i64,i64,i64,i64,i64,str,i64,str,i64,str,str,i64,str,i64,i64,str,str,str,i64,i64,i64,i64,i64,i64,str,str,str,i64,i64,i64,str,str,i64
1,60,"""RL""",65,8450,"""Pave""",null,"""Reg""","""Lvl""","""AllPub""","""Inside""","""Gtl""","""CollgCr""","""Norm""","""Norm""","""1Fam""","""2Story""",7,5,2003,2003,"""Gable""","""CompShg""","""VinylSd""","""VinylSd""","""BrkFace""",196,"""Gd""","""TA""","""PConc""","""Gd""","""TA""","""No""","""GLQ""",706,"""Unf""",0,…,854,0,1710,1,0,2,1,3,1,"""Gd""",8,"""Typ""",0,null,"""Attchd""",2003,"""RFn""",2,548,"""TA""","""TA""","""Y""",0,61,0,0,0,0,null,null,null,0,2,2008,"""WD""","""Normal""",208500
2,20,"""RL""",80,9600,"""Pave""",null,"""Reg""","""Lvl""","""AllPub""","""FR2""","""Gtl""","""Veenker""","""Feedr""","""Norm""","""1Fam""","""1Story""",6,8,1976,1976,"""Gable""","""CompShg""","""MetalSd""","""MetalSd""",null,0,"""TA""","""TA""","""CBlock""","""Gd""","""TA""","""Gd""","""ALQ""",978,"""Unf""",0,…,0,0,1262,0,1,2,0,3,1,"""TA""",6,"""Typ""",1,"""TA""","""Attchd""",1976,"""RFn""",2,460,"""TA""","""TA""","""Y""",298,0,0,0,0,0,null,null,null,0,5,2007,"""WD""","""Normal""",181500
3,60,"""RL""",68,11250,"""Pave""",null,"""IR1""","""Lvl""","""AllPub""","""Inside""","""Gtl""","""CollgCr""","""Norm""","""Norm""","""1Fam""","""2Story""",7,5,2001,2002,"""Gable""","""CompShg""","""VinylSd""","""VinylSd""","""BrkFace""",162,"""Gd""","""TA""","""PConc""","""Gd""","""TA""","""Mn""","""GLQ""",486,"""Unf""",0,…,866,0,1786,1,0,2,1,3,1,"""Gd""",6,"""Typ""",1,"""TA""","""Attchd""",2001,"""RFn""",2,608,"""TA""","""TA""","""Y""",0,42,0,0,0,0,null,null,null,0,9,2008,"""WD""","""Normal""",223500
4,70,"""RL""",60,9550,"""Pave""",null,"""IR1""","""Lvl""","""AllPub""","""Corner""","""Gtl""","""Crawfor""","""Norm""","""Norm""","""1Fam""","""2Story""",7,5,1915,1970,"""Gable""","""CompShg""","""Wd Sdng""","""Wd Shng""",null,0,"""TA""","""TA""","""BrkTil""","""TA""","""Gd""","""No""","""ALQ""",216,"""Unf""",0,…,756,0,1717,1,0,1,0,3,1,"""Gd""",7,"""Typ""",1,"""Gd""","""Detchd""",1998,"""Unf""",3,642,"""TA""","""TA""","""Y""",0,35,272,0,0,0,null,null,null,0,2,2006,"""WD""","""Abnorml""",140000
5,60,"""RL""",84,14260,"""Pave""",null,"""IR1""","""Lvl""","""AllPub""","""FR2""","""Gtl""","""NoRidge""","""Norm""","""Norm""","""1Fam""","""2Story""",8,5,2000,2000,"""Gable""","""CompShg""","""VinylSd""","""VinylSd""","""BrkFace""",350,"""Gd""","""TA""","""PConc""","""Gd""","""TA""","""Av""","""GLQ""",655,"""Unf""",0,…,1053,0,2198,1,0,2,1,4,1,"""Gd""",9,"""Typ""",1,"""TA""","""Attchd""",2000,"""RFn""",3,836,"""TA""","""TA""","""Y""",192,84,0,0,0,0,null,null,null,0,12,2008,"""WD""","""Normal""",250000
6,50,"""RL""",85,14115,"""Pave""",null,"""IR1""","""Lvl""","""AllPub""","""Inside""","""Gtl""","""Mitchel""","""Norm""","""Norm""","""1Fam""","""1.5Fin""",5,5,1993,1995,"""Gable""","""CompShg""","""VinylSd""","""VinylSd""",null,0,"""TA""","""TA""","""Wood""","""Gd""","""TA""","""No""","""GLQ""",732,"""Unf""",0,…,566,0,1362,1,0,1,1,1,1,"""TA""",5,"""Typ""",0,null,"""Attchd""",1993,"""Unf""",2,480,"""TA"""

In [3]:
houses = raw.clone()
print(f"Shape: {houses.shape}")
print(f"Columns: {houses.columns}")
houses.schema

Shape: (1460, 81)
Columns: ['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd', 'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1', 'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating', 'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual', 'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType', 'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual', 'GarageCond', 'PavedDrive', 'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'ScreenP

Schema([('Id', Int64),
        ('MSSubClass', Int64),
        ('MSZoning', String),
        ('LotFrontage', Int64),
        ('LotArea', Int64),
        ('Street', String),
        ('Alley', String),
        ('LotShape', String),
        ('LandContour', String),
        ('Utilities', String),
        ('LotConfig', String),
        ('LandSlope', String),
        ('Neighborhood', String),
        ('Condition1', String),
        ('Condition2', String),
        ('BldgType', String),
        ('HouseStyle', String),
        ('OverallQual', Int64),
        ('OverallCond', Int64),
        ('YearBuilt', Int64),
        ('YearRemodAdd', Int64),
        ('RoofStyle', String),
        ('RoofMatl', String),
        ('Exterior1st', String),
        ('Exterior2nd', String),
        ('MasVnrType', String),
        ('MasVnrArea', Int64),
        ('ExterQual', String),
        ('ExterCond', String),
        ('Foundation', String),
        ('BsmtQual', String),
        ('BsmtCond', String),
        ('Bsmt

Total of columns : 81
Total of rows : 1460
Data quantity : more or less 115k

The dataframe seems to contain a lot of extremely detailed categories regarding various types of properties and amenities.

We will have to find which may have an impact on the sale prices of houses.

### Dealing with missing values

Before going any further, we will check if there are missing values in this dataframe and we will deal with these missing values.

Some columns seem to have null data. But when we look at the data description file given with the dataset, it seems that some data with the entry "NA" have been transformed to "NaN" in the dataframe.

For example, in the data description file, the column "Alley" should have the following data : "NA" for "No alley access". Instead it was replaced by "NaN" for null.

We will have a look at these columns to make the appropriate modifications.

In [4]:
# Visualize missing values
nulls = houses.null_count().unpivot().filter(pl.col("value") > 0).sort("value", descending=True)
fig = px.bar(
    nulls.to_pandas(), x="variable", y="value",
    color_discrete_sequence=[PASTEL[4]],
    labels={"variable": "Column", "value": "Missing count"},
    title="Missing Values per Column", text_auto=True,
)
fig.update_layout(showlegend=False, width=900, height=400)
fig.show()

In [5]:
# Columns where NA means "None/No feature" per data description
na_means_none = [
    "Alley", "BsmtQual", "BsmtCond", "BsmtExposure", "BsmtFinType1",
    "BsmtFinType2", "FireplaceQu", "GarageType", "GarageFinish",
    "GarageQual", "GarageCond", "PoolQC", "Fence", "MiscFeature",
]
houses = houses.with_columns(
    pl.col(c).fill_null("NA") for c in na_means_none if c in houses.columns
)

# LotFrontage: fill with mode
lot_mode = houses["LotFrontage"].drop_nulls().mode().item()
houses = houses.with_columns(pl.col("LotFrontage").fill_null(lot_mode))

# MasVnrType & MasVnrArea: fill with mode
houses = houses.with_columns(
    pl.col("MasVnrType").fill_null(houses["MasVnrType"].drop_nulls().mode().item()),
    pl.col("MasVnrArea").fill_null(houses["MasVnrArea"].drop_nulls().mode().item()),
)

# Electrical: single null — fill with mode
houses = houses.with_columns(
    pl.col("Electrical").fill_null(houses["Electrical"].drop_nulls().mode().item())
)

# GarageYrBlt: fill with YearBuilt (no garage → assume built with the house)
houses = houses.with_columns(
    pl.col("GarageYrBlt").fill_null(pl.col("YearBuilt"))
)

# Verify no nulls remain
remaining = houses.null_count().unpivot().filter(pl.col("value") > 0)
print(f"Remaining nulls: {remaining.height} columns" if remaining.height else "No nulls remain.")

No nulls remain.


### What is in the column Sale Prices?

Let's have a look at the content of the column "SalePrice" to see how sale prices are distributed in our dataframe.

In [6]:
fig = px.histogram(
    houses.to_pandas(), x="SalePrice", nbins=50,
    color_discrete_sequence=[PASTEL[2]],
    title="Sales Price Distribution",
    marginal="box",
)
fig.update_layout(width=800, height=450)
fig.show()

The distribution is uneven. Most data is for houses sold between $100000 and $200000.

### Correlation matrix between columns with numerical values

Now, let's start building a correlation matrix to have a look at which data might have an impact on the sales price.

First, we will store the correlation data into a separate dataset and then we will filter to keep only potentially relevant correlations. Otherwise it would simply be unreadable due to the massive amount of data at our disposal.

In [7]:
# Numeric correlation with SalePrice (sorted)
numeric_cols = [c for c in houses.columns if houses[c].dtype in (pl.Float64, pl.Int64) and c != "Id"]
corr_sale = houses.select(numeric_cols).to_pandas().corr()["SalePrice"].drop("SalePrice").sort_values(ascending=False)

fig = px.bar(
    x=corr_sale.values, y=corr_sale.index,
    orientation="h", color=corr_sale.values,
    color_continuous_scale="RdBu_r", color_continuous_midpoint=0,
    title="Numeric Feature Correlation with SalePrice",
    labels={"x": "Correlation", "y": "Feature"},
)
fig.update_layout(height=800, width=700, yaxis=dict(autorange="reversed"))
fig.show()

#### How does it work?

The values are between -1 and +1.

A positive correlation is observed when the values of the two variables being compared increase together.

And a negative correlation when the values of one variable increase while the values of the other variable decrease.

At first glance, some values seem to have a high correlation with the sale price of a house:
- OverallQual has a 0.79 correlation with SalePrice
- GrLivArea has a 0.71 correlation with SalePrice
- GarageCars has a 0.64 correlation with SalePrice

And so on.

Also, some seem to have almost no impact on the price of a house : all those values where skipped because we requested for our correlations to be superior to 0.20 or inferior to -0.20. So all the columns that were not included in the correlation matrix could be considered to have a limited or no positive or negative impact on the price of a house.

In [21]:
# Full correlation heatmap (top features only, above |0.4|)
top_features = corr_sale[corr_sale.abs() >= 0.4].index.tolist() + ["SalePrice"]
corr_top = houses.select(top_features).to_pandas().corr()

fig = px.imshow(
    corr_top, text_auto=".2f", color_continuous_scale="RdBu_r",
    zmin=-1, zmax=1, title="Correlation Matrix (|r| ≥ 0.4 with SalePrice)",
)
fig.update_layout(width=800, height=700)
fig.show()

We can consider the columns above 0.40 to have an impact on the sales price of a house. These columns are :
- OverallQual for the overall quality of the house
- GrLivArea for the above ground living area measured in square feet
- GarageCars for the size of the garage measured in car capacity
- GarageArea for the size of the garage measured in square feet
- FullBath for the number of bathrooms above ground
- TotalBsmtSF for the total square feet of the basement area
- 1stFlrSF for the size of the first floor measured in square feet
- TotRmsAbvGrd for the total number of rooms above ground (excluding bathrooms)
- YearBuilt for the year in which the house was built
- YearRemodAdd for remodel date
- MasVnrArea for the masonry veneer area in square feet
- Fireplaces for the number of fireplaces in the house

A few notes:
- GarageCars and GarageArea are pretty similar, they represent the size of the garage but are simply measured differently : one is measured with square feet, the other is measured with the number of cars that could fit in the garage.
- Most of these represent sizes in square feet which could mean that the size does count !
- All of the columns used have numerical data, categorical data was not included in these correlations.

### Correlation with encoded categorical values

Now, to be a bit more thorough, we will transform categorical data using ordinal encoding in order to add them to our correlation analysis.

In [9]:
# Ordinal-encode all string columns for correlation analysis
houses_encoded = houses.drop("Id").clone()
for col_name in houses_encoded.columns:
    if houses_encoded[col_name].dtype == pl.String:
        cats = sorted(houses_encoded[col_name].unique().to_list())
        mapping = {v: i for i, v in enumerate(cats)}
        houses_encoded = houses_encoded.with_columns(
            pl.col(col_name).replace_strict(mapping, return_dtype=pl.Int64)
        )

# Full correlation with SalePrice including encoded categoricals
corr_full = houses_encoded.to_pandas().corr()["SalePrice"].drop("SalePrice").sort_values(ascending=False)

fig = px.bar(
    x=corr_full.values, y=corr_full.index,
    orientation="h", color=corr_full.values,
    color_continuous_scale="RdBu_r", color_continuous_midpoint=0,
    title="All Features Correlation with SalePrice (incl. encoded categoricals)",
    labels={"x": "Correlation", "y": "Feature"},
)
fig.update_layout(height=1000, width=700, yaxis=dict(autorange="reversed"))
fig.show()

A few notes :
- The above 0.40 ranking did not change.
- The below or equal to -0.40 ranking did change : some of our categorical data is shown as having a correlation on the sales price of the house.

The columns that are below or equal to -0.40 are, starting from the lowest of all :
- ExterQual for the quality of the material on the exterior of the house
- BsmtQual for the evaluation of the height of the basement
- KitchenQual for the evaluation of the kitchen quality
- GarageFinish for the quality of the interior finish of the garage
- GarageType for the type of the garage and its location compared to the house
- HeatingQC for the heating quality and condition

These columns seem to be linked to quality and the evaluation of quality for certain aspects of a house. I wonder if we could link these elements to the column "OverallQual" (for overall quality) which sits at the top of the correlation heatmap.

### Correlation with the overall quality of a house

In [10]:
corr_quality = houses_encoded.to_pandas().corr()["OverallQual"].drop("OverallQual").sort_values(ascending=False)

fig = px.bar(
    x=corr_quality.values, y=corr_quality.index,
    orientation="h", color=corr_quality.values,
    color_continuous_scale="RdBu_r", color_continuous_midpoint=0,
    title="All Features Correlation with OverallQual",
    labels={"x": "Correlation", "y": "Feature"},
)
fig.update_layout(height=1000, width=700, yaxis=dict(autorange="reversed"))
fig.show()

This seems to confirm that there is a correlation between the column for overall quality and the columns "ExterQual", "BsmtQual", "KitchenQual", "GarageType", "HeatingQC" and "GarageFinish".

### The Case of the Neighborhood

I have to admit that I'm a bit surprised that the column "Neighborhood" has a correlation of 0.21 with the "SalePrice" column. We would tend to think that the neighborhood would have an influence, positive or negative, on the price. We will test that theory.

In [11]:
# Sales count per neighborhood
neigh_counts = houses["Neighborhood"].value_counts().sort("count", descending=True)
fig = px.bar(
    neigh_counts.to_pandas(), x="Neighborhood", y="count",
    color_discrete_sequence=[PASTEL[0]],
    title="Number of Sales per Neighborhood", text_auto=True,
)
fig.update_layout(width=900, height=400)
fig.show()

In [12]:
# Neighborhood impact on price — box plot (much better than strip plot for distributions)
neigh_order = (
    houses.group_by("Neighborhood")
    .agg(pl.col("SalePrice").median().alias("median_price"))
    .sort("median_price", descending=True)
)["Neighborhood"].to_list()

fig = px.box(
    houses.to_pandas(), x="Neighborhood", y="SalePrice",
    color_discrete_sequence=[PASTEL[3]],
    title="SalePrice Distribution by Neighborhood",
    category_orders={"Neighborhood": neigh_order},
)
fig.update_layout(width=1000, height=500)
fig.show()

There is a real discrepancy between the highest and lowest neighborhoods in term of mean and median amount of dollars required to purchase a house. This confirms that the neighborhood indeed has an impact on the sales price.

### The case of the Month and Year

Although MoSold and YrSold have a low correlation with SalePrice, I'm curious to test if there is a link between these elements and the sales price.

In [13]:
# Sales count per month (with year breakdown)
fig = px.histogram(
    houses.to_pandas(), x="MoSold", color="YrSold",
    barmode="group", color_discrete_sequence=PASTEL,
    title="Sales per Month per Year",
    nbins=12,
)
fig.update_layout(width=800, height=400, bargap=0.1)
fig.show()

In [14]:
# SalePrice by month — box plot
fig = px.box(
    houses.to_pandas(), x="MoSold", y="SalePrice",
    color_discrete_sequence=[PASTEL[5]],
    title="SalePrice by Month Sold",
)
fig.update_layout(width=800, height=400)
fig.show()

Sales appear to happen most around May - June - July, which is not that shocking.

Confirmed : the month and year do not have an impact on the sales price.

However, the years we have here are between 2006 and 2010. It seems logical that on a short time frame prices do not rise that much. It would be another story if we had data from 1950 to today.

Adding the year does change the repartition, except that we seem to be missing the data for 2010 from June/July and the rest of the year. That could be explained by the date when this dataset was first compiled.

### Category group comparisons

It may be interesting to put some columns in similar categories and compare their correlation with SalePrice.

In [15]:
# Define column groups
groups = {
    "Lot & Sale": [
        "Alley", "Condition1", "Condition2", "LandContour", "LotArea", "LotConfig",
        "LotFrontage", "LotShape", "MSSubClass", "MSZoning", "Neighborhood", "Street",
        "Utilities", "MiscFeature", "MiscVal", "MoSold", "YrSold", "SaleType", "SaleCondition",
    ],
    "Building Size & Features": [
        "GrLivArea", "1stFlrSF", "2ndFlrSF", "LowQualFinSF",
        "CentralAir", "Electrical", "Fireplaces", "FireplaceQu",
        "Functional", "Heating", "HeatingQC",
    ],
    "Building Details": [
        "BldgType", "Exterior1st", "Exterior2nd", "ExterQual", "ExterCond",
        "Foundation", "HouseStyle", "MasVnrType", "MasVnrArea",
        "OverallQual", "OverallCond", "RoofStyle", "RoofMatl", "YearBuilt", "YearRemodAdd",
    ],
    "Interior Spaces": [
        "BsmtCond", "BsmtExposure", "BsmtFinType1", "BsmtFinSF1",
        "BsmtFinType2", "BsmtFinSF2", "BsmtQual", "BsmtUnfSF", "TotalBsmtSF",
        "BedroomAbvGr", "BsmtFullBath", "BsmtHalfBath", "FullBath",
        "HalfBath", "KitchenAbvGr", "KitchenQual", "TotRmsAbvGrd",
    ],
    "Garage & Exterior": [
        "GarageArea", "GarageCars", "GarageCond", "GarageFinish",
        "GarageQual", "GarageType", "GarageYrBlt",
        "EnclosedPorch", "Fence", "LandSlope", "OpenPorchSF",
        "PavedDrive", "PoolArea", "PoolQC", "ScreenPorch", "WoodDeckSF", "3SsnPorch",
    ],
}

# Show per-group correlation heatmaps with SalePrice
for group_name, cols in groups.items():
    valid = [c for c in cols if c in houses_encoded.columns] + ["SalePrice"]
    grp_corr = houses_encoded.select(valid).to_pandas().corr()["SalePrice"].drop("SalePrice").sort_values(ascending=False)
    fig = px.bar(
        x=grp_corr.values, y=grp_corr.index,
        orientation="h", color=grp_corr.values,
        color_continuous_scale="RdBu_r", color_continuous_midpoint=0,
        title=f"{group_name} — Correlation with SalePrice",
        labels={"x": "Correlation", "y": "Feature"},
    )
    fig.update_layout(height=max(300, len(grp_corr) * 25), width=700, yaxis=dict(autorange="reversed"))
    fig.show()

### Top feature scatter plots vs SalePrice

In [16]:
# Scatter plots: top 6 correlated features vs SalePrice
top6 = ["OverallQual", "GrLivArea", "GarageCars", "GarageArea", "TotalBsmtSF", "1stFlrSF"]

fig = make_subplots(rows=2, cols=3, subplot_titles=top6)
for i, feat in enumerate(top6):
    r, c = divmod(i, 3)
    fig.add_trace(go.Scatter(
        x=houses[feat].to_list(), y=houses["SalePrice"].to_list(),
        mode="markers", marker=dict(color=PASTEL[i], opacity=0.5, size=4),
        showlegend=False,
    ), row=r + 1, col=c + 1)
    fig.update_xaxes(title_text=feat, row=r + 1, col=c + 1)
fig.update_yaxes(title_text="SalePrice", col=1)
fig.update_layout(title="Top Features vs SalePrice", height=600, width=1000)
fig.show()

### Summary of low-impact columns

Here is a complete list of the columns that are considered to have a low impact on the price (correlation btw -0.15 and 0.15):

- Alley, Street, Condition1, Condition2, MoSold, YrSold, SaleType
- LandContour, LotConfig, Utilities, MiscFeature, MiscVal, MSSubClass, MSZoning
- 3SsnPorch, EnclosedPorch, Fence, LandSlope, PoolArea, PoolQC, ScreenPorch
- Functional, LowQualFinSF, FireplaceQu, Heating
- RoofMatl, ExterCond, Exterior2nd, Exterior1st, MasVnrType, OverallCond, BldgType
- BsmtCond, BsmtFinType2, BsmtFinSF2, BsmtHalfBath, BsmtFinType1, KitchenAbvGr

In [17]:
# Programmatic list of low-impact columns
low_impact = corr_full[corr_full.abs() < 0.15].index.tolist()
print(f"{len(low_impact)} columns with |correlation| < 0.15 with SalePrice:")
print(low_impact)

36 columns with |correlation| < 0.15 with SalePrice:
['Fence', 'RoofMatl', 'ExterCond', 'Functional', 'ScreenPorch', 'Exterior2nd', 'Exterior1st', 'PoolArea', 'BsmtCond', 'Condition1', 'Alley', 'BsmtFinType2', 'LandSlope', 'MoSold', '3SsnPorch', 'Street', 'LandContour', 'Condition2', 'BsmtFinSF2', 'Utilities', 'BsmtHalfBath', 'MiscVal', 'LowQualFinSF', 'YrSold', 'SaleType', 'MiscFeature', 'LotConfig', 'OverallCond', 'MSSubClass', 'BldgType', 'FireplaceQu', 'BsmtFinType1', 'Heating', 'PoolQC', 'EnclosedPorch', 'KitchenAbvGr']


### End of analysis

- For the machine learning prediction : it might be interesting for the machine learning prediction to start using all columns in an almost clean/untouched dataframe - well except for missing values - and then later use another dataframe which excludes all the columns with low impact that we identified before that.
- For further data analysis : it could be interesting to explore more hypothesis and test more theories on this dataset.